packages/core/content-manager/admin/src/pages/EditView/components/EditorToolbarObserver.tsx

```diff
@@ -0,0 +1,144 @@
+import * as React from 'react';
+
+import { useElementOnScreen } from '@strapi/admin/strapi-admin';
+import { IconButton, Menu } from '@strapi/design-system';
+import { More } from '@strapi/icons';
+import { useIntl } from 'react-intl';
+
+/* -------------------------------------------------------------------------------------------------
+ * ObservedToolbarComponent
+ * -----------------------------------------------------------------------------------------------*/
+
+interface ObservedToolbarComponentProps {
+  index: number;
+  lastVisibleIndex: number;
+  setLastVisibleIndex: React.Dispatch<React.SetStateAction<number>>;
+  rootRef: React.RefObject<HTMLElement>;
+  children: React.ReactNode;
+}
+
+const ObservedToolbarComponent = ({
+  index,
+  lastVisibleIndex,
+  setLastVisibleIndex,
+  rootRef,
+  children,
+}: ObservedToolbarComponentProps) => {
+  const isVisible = index <= lastVisibleIndex;
+
+  const containerRef = useElementOnScreen<HTMLDivElement>(
+    (isVisible) => {
+      /**
+       * It's the MoreMenu's job to make an item not visible when there's not room for it.
+       * But we need to react here to the element becoming visible again.
+       */
+      if (isVisible) {
+        setLastVisibleIndex((prev) => Math.max(prev, index));
+      }
+    },
+    { threshold: 1, root: rootRef.current }
+  );
+
+  return (
+    <div
+      ref={containerRef}
+      style={{
+        /**
+         * Use visibility so that the element occupies the space if requires even when there's not
+         * enough room for it to be visible. The empty reserved space will be clipped by the
+         * overflow:hidden rule on the parent, so it doesn't affect the UI.
+         * This way we can keep observing its visiblity and react to browser resize events.
+         */
+        visibility: isVisible ? 'visible' : 'hidden',
+      }}
+    >
+      {children}
+    </div>
+  );
+};
+
+/* -------------------------------------------------------------------------------------------------
+ * EditorToolbarObserver
+ * -----------------------------------------------------------------------------------------------*/
+
+export interface ObservedComponent {
+  toolbar: React.ReactNode;
+  menu: React.ReactNode;
+  key: string;
+}
+
+export const EditorToolbarObserver = ({
+  observedComponents,
+  editor = 'blocks',
+}: {
+  observedComponents: ObservedComponent[];
+  editor: 'blocks' | 'markdown';
```

Comment:
This editor prop is only used for the menu trigger variant right?

If so I have a slight preference for a triggerVariant prop over this. I feel a bit weird about an abstraction needing to specific from where it's being called (although yes that is exactly what I did with my location prop 🙃)